### PR TITLE
fix: support link in publication that don't have http or https

### LINKED ISSUE #3490: Support link in publication that don't have http or https
https://lenster.xyz/posts/0x01d7ea-0x02

<sub>[ENG-18](https://linear.app/heyxyz/issue/ENG-18/support-link-in-publication-that-dont-have-http-or-https)</sub>

### PR DESCRIPTION
**Pull request by Hitesh-sisara**
## What does this PR do?

test vid : https://www.loom.com/share/e7d7950b0e5643e19d3178c55478ac1e?sid=5351546c-45a8-4fbc-ace4-82e6e66d6e2c

### <samp>🤖 Generated by Copilot at 9f18ae8</samp>

Improved `ExternalLink` component to handle invalid or incomplete links. Added protocol validation and defaulting logic and prevented rendering of non-linkable strings.

Fixes #3490
/claim #3490

- [x] Enhancement (non-breaking small changes to existing functionality)

### <samp>🤖 Generated by Copilot at 9f18ae8</samp>

* Improve error handling and default protocol for `ExternalLink` component ([link](https://github.com/heyxyz/hey/pull/3869/files?diff=unified&w=0#diff-6dc897ab1139883d81982c4d42b94e4daed0ea6466ca5fb404da7e71462bc866L11-R18))

## Emoji

🛠️🔗🙅‍♂️

### CODE REVIEW

**Comment by bigint** [comment]
@Hitesh-sisara should we close #3864 ?

**Review by bigint — CHANGES_REQUESTED** [review]
thanks for the PR!

this gonna make unknown tlds as url?

eg:

what are you `doing.hitesh` i'm good
you are `great.but` need to grow ?

**Comment by foolo** [comment]
@bigint IMO this issue needs unit tests, because it's such a complex problem, with a lot of caveats and corner cases, and it's really easy to break one thing while fixing another. If we just do small hotfixes without test, it will be a never ending story of pull requests.

**Comment by bigint** [comment]
Yup +1 we need unit tests for this for sure 🙇🏼 lmao 😓

**Comment by Hitesh-sisara** [comment]
> thanks for the PR!
> 
> this gonna make unknown tlds as url?
> 
> eg:
> 
> what are you `doing.hitesh` i'm good you are `great.but` need to grow ?

if the link contains only one segment like (abc.xyz), it won't be considered as a URL to avoid false positives. but if is contains (abc.xyz/a) then it consider as link

**Comment by Hitesh-sisara** [comment]
> > thanks for the PR!
> > this gonna make unknown tlds as url?
> > eg:
> > what are you `doing.hitesh` i'm good you are `great.but` need to grow ?
> 
> if the link contains only one segment like (abc.xyz), it won't be considered as a link to avoid false positives. but if is contains (abc.xyz/a) then it consider as link

@bigint which way should i implement the current way is file or we should check for 

> > if the link contains only one segment like (abc.xyz), it won't be considered as a link to avoid false positives. but if is contains (abc.xyz/a) then it consider as link
> 
> but even single segment should be considered as a link, if we support one and not other will create inconsistency 🙇🏼

ok then i should implement logic to check for all tlds

**Comment by Hitesh-sisara** [comment]
@bigint can you guide how to login in to dev mode on localhost currently it ask for lens access

**Comment by Hitesh-sisara** [comment]
> @Hitesh-sisara should we close #3864 ?

Yes

**Comment by Hitesh-sisara** [comment]
@bigint  check : https://www.loom.com/share/e7d7950b0e5643e19d3178c55478ac1e?sid=5351546c-45a8-4fbc-ace4-82e6e66d6e2c

**Comment by Hitesh-sisara** [comment]
I could have included TLD checking directly in the regex, but the list of TLDs is quite extensive, and verifying them for every piece of text can slow down performance. So, the regex will identify links, and then a separate check will be done to verify if the link contains valid TLD.

**Comment by foolo** [comment]
> I could have included TLD checking directly in the regex, but the list of TLDs is quite extensive, and verifying them for every piece of text can slow down performance. So, the regex will identify links, and then a separate check will be done to verify if the link contains valid TLD.

@Hitesh-sisara 👍 I think this is a good approach. Before, we had a regex that was more than 8000 characters long :) and it caused a crash. ( https://github.com/heyxyz/hey/issues/2183 ) A simple regex followed by TLD logic is a good hybrid solution.

**Comment by Hitesh-sisara** [comment]
@bigint take a look

File: apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx [hunk 1]
```diff
@@ -1,21 +1,27 @@
-import stopEventPropagation from '@hey/lib/stopEventPropagation';
 import type { MarkupLinkProps } from '@hey/types/misc';
 import Link from 'next/link';
 import type { FC } from 'react';
 
+import { isUrlContainsValidTld } from '../../../../../../../packages/data/utils/check_valid_tld';
```
Comment (bigint) [inline apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx:5]: use proper monorepo imports

File: packages/data/utils/tlds.ts
Comment (bigint) [inline packages/data/utils/tlds.ts:1]: we can use dependencies for this, instead of using beefy file 🙈

**Review by bigint — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx (hunk 1 above)
Comment (Hitesh-sisara) [reply apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx]: fixed

File: packages/data/utils/tlds.ts
Comment (Hitesh-sisara) [reply packages/data/utils/tlds.ts:1]: used tlds package i

**Comment by Hitesh-sisara** [comment]
@bigint done suggested changes

File: apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import stopEventPropagation from '@hey/lib/stopEventPropagation';
+import { isUrlContainsValidTld } from '@hey/data/utils/check_valid_tld';
```
Comment (bigint) [inline apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx:1]: move this to lib folder, data folder are all static files and values

File: apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx [hunk 3]
```diff
@@ -8,14 +8,19 @@ const ExternalLink: FC<MarkupLinkProps> = ({ href, title = href }) => {
     return null;
   }
 
+  if (!isUrlContainsValidTld(href)) {
+    // if href not contains valid tld then we will not render as link
+
+    return href;
+  }
```
Comment (bigint) [inline apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx:15]:
```suggestion
  if (!isUrlContainsValidTld(href)) {
    return href;
  }
```

File: packages/data/tests/regex/tld.spec.ts
Comment (bigint) [inline packages/data/tests/regex/tld.spec.ts:1]: move this to lib as well

**Comment by bigint** [comment]
https://web-git-fork-hitesh-sisara-fix-link-heyxyz.vercel.app/posts/0x24b6-0x24-DA-52bc214f

Screenshot 2023-10-07 at 2 42 36 PM

http url looks odd

**Review by bigint — CHANGES_REQUESTED** [review]

File: packages/data/tests/regex/tld.spec.ts
Comment (Hitesh-sisara) [reply packages/data/tests/regex/tld.spec.ts:1]: moved to lib

File: apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx (hunk 2 above)
Comment (Hitesh-sisara) [reply apps/web/src/components/Shared/Markup/MarkupLink/ExternalLink.tsx]: moved to lib

**Comment by Hitesh-sisara** [comment]
> https://web-git-fork-hitesh-sisara-fix-link-heyxyz.vercel.app/posts/0x24b6-0x24-DA-52bc214f
> 
> Screenshot 2023-10-07 at 2 42 36 PM
> http url looks odd

this i am not able to replicate in my testing working as expected,   can you recheck there were no space after http ? -- "http:// tape.xyz"

**Comment by bigint** [comment]
I just tried `http://example.com`

**Comment by Hitesh-sisara** [comment]
> I just tried `http://example.com`

try once more

**Comment by bigint** [comment]
https://github.com/heyxyz/hey/assets/69431456/601fba9f-b38b-4476-bd39-2e94161b8d31

oops something broke too while checking!

https://web-git-fork-hitesh-sisara-fix-link-heyxyz.vercel.app/posts/0x24b6-0x24-DA-52bc214f

**Comment by Hitesh-sisara** [comment]
> Screen.Recording.2023-10-08.at.10.12.43.AM.mov 
> 
> oops something broke too while checking!
> 
> https://web-git-fork-hitesh-sisara-fix-link-heyxyz.vercel.app/posts/0x24b6-0x24-DA-52bc214f

fixed : error was happening while getNft.ts trying to parse URLs

**Comment by bigint** [comment]
Link T-668

**Comment by bigint** [comment]
We need to wait till https://github.com/heyxyz/hey/pull/3892/ goes out 🙇🏼

**Comment by Hitesh-sisara** [comment]
> We need to wait till #3892 goes out 🙇🏼

ok do it earliest

**Comment by bigint** [comment]
@Hitesh-sisara is everything working as expected?

**Comment by bigint** [comment]
Closing this, we can rethink this later!
